_stbt/core.py

```diff
@@ -1213,43 +1217,67 @@ def wait_until(callable_, timeout_secs=10, interval_secs=0, stable_secs=0):
         if not wait_until(lambda: match("xyz.png")):
             do_something_else()
 
-    There are some drawbacks to using `assert` instead of `wait_for_match`:
-
-    * The exception message won't contain the reason why the match failed
-      (unless you specify it as a second parameter to `assert`, which is
-      tedious and we don't expect you to do it), and
-    * The exception won't have the offending video-frame attached.
-
-    We hope to solve both of the above drawbacks at some point in the future.
-
-    ``wait_until`` was added in stb-tester v22. The ``stable_secs`` parameter
-    was added in v28.
+        # Wait for a menu selection to change. Here ``Menu`` is a `FrameObject`
+        # with a property called `selection` that returns a string with the
+        # name of the currently-selected menu item:
+        # The return value (``menu``) is an instance of ``Menu``.
+        menu = wait_until(Menu, predicate=lambda x: x.selection == "Home")
+
+        # Wait for a match to stabilise position, returning the first stable
+        # match. Used in performance measurements, for example to wait for a
+        # selection highlight to finish moving:
+        press("KEY_DOWN")
+        start_time = time.time()
+        match_result = wait_until(lambda: stbt.match("selection.png"),
+                                  predicate=lambda x: x and x.region,
+                                  stable_secs=2)
+        assert match_result
+        end_time = match_result.time  # this is the first stable frame
+        print "Transition took %s seconds" % (end_time - start_time)
+
+    ``wait_until`` was added in stb-tester v22. The ``predicate``,
+    ``stable_secs``, and ``ignored_exceptions`` parameters were added in v28.
     """
     import time
 
+    if predicate is None:
+        predicate = lambda x: x
     stable_value = None
+    stable_predicate_value = None
     expiry_time = time.time() + timeout_secs
 
     while True:
         t = time.time()
-        value = callable_()
+        try:
+            value = callable_()
+        except ignored_exceptions as e:
+            debug("wait_until: Ignoring exception %r" % e)
```

Comment:
Need to test unicode behaviour here? We've had trouble with printing exceptions in the past.

Reply:
This comment was before we dropped support for Python 2.